Basic driver management for Appium 2.0

jlipps opened the pull request:
Initial work on Appium 2.0. Taking over from https://github.com/appium/appium/pull/13461. When merged, this will merely go into the appium/appium `2.0` branch (not master).

jlipps committed:
breaking change: do not depend directly on drivers anymore
instead look for a reference to them in a drivers json file which will be
managed by the appium driver cli tool. this is a WIP and tests are not yet fixed

(breaking change)

feat: add driver CLI and subcommands; start to implement them beginning with list

feat: enable showing possible updates during 'driver list'

feat: enable installation of drivers via 'driver install'

feat: enable installation of drivers via local, git, and github specs

fix: lint and unit test updates to get everything passing

fix: make sure travis installs fakedriver

refactor: move cli helpers into own file and clean up driver json validation

jlipps commented:
@mykola-mokhnach wanted to give you a chance to review the initial code as well.

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,157 @@
+  } catch (err) {
+    throw new Error(`Appium had trouble loading the driver installation ` +
+                    `cache file (${DRIVERS_JSON}). Ensure it exists and is ` +
+                    `readable. Specific error: ${err}`);
```
err.message

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,157 @@
+  for (const driverName of Object.keys(INSTALLED_DRIVERS)) {
+    const platformNames = INSTALLED_DRIVERS[driverName].platformNames;
+    if (!_.isArray(platformNames)) {
+      msg += `Drivers.json entry for '${driverName}' has missing or ` +
```
file name can be a constant

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,157 @@
+
+function printAvailableDrivers () {
+  const driverNames = Object.keys(INSTALLED_DRIVERS);
+  if (!driverNames.length) {
```
_.isEmpty

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,157 @@
+
+    if (aNameMatches && pNameMatches) {
+      return {driverName, ...installed};
+    } else if (aNameMatches) {
```
else is redundant

jlipps replied:
👍

jlipps replied:
actually, it's not redundant. it's handling the case where `aNameMatches && !pNameMatches`

mykola-mokhnach replied:
you can keep `if (aNameMatches)`. I was only talking about `else`

jlipps replied:
oh i see 👍

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,157 @@
+    const installed = INSTALLED_DRIVERS[driverName];
+    const aNameMatches = installed.automationName.toLowerCase() ===
+                         automationName.toLowerCase();
+    const lower = (x) => x.toLowerCase();
```
use _.toLower()

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,157 @@
+function getDriverBySupport (automationName, platformName) {
+  for (const driverName of Object.keys(INSTALLED_DRIVERS)) {
+    const installed = INSTALLED_DRIVERS[driverName];
+    const aNameMatches = installed.automationName.toLowerCase() ===
```
is automationName always set?

jlipps replied:
Yes, we error out earlier in the flow if it's not set.

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,157 @@
+           `'${automationName}' and platformName '${platformName}'`);
+
+  try {
+    const {driverName, mainClass, pkg, version} = getDriverBySupport(automationName, platformName);
```
what the function name means (BySupport) ?

jlipps replied:
get the driver based on its supported automation name and platform names. kinda long-winded :-)

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,157 @@
+                `'${automationName}' and platformName ${platformName}'. ` +
+                `Have you installed a driver that supports those ` +
+                `capabilities? Run 'appium driver list --installed' to see. ` +
+                `(Lower-level error: ${err})`;
```
err.message

mykola-mokhnach commented on lib/main.js:
```diff
@@ -90,6 +93,19 @@ function logServerPort (address, port) {
+  }
+
+  if (!_.includes(['driver', 'server', '-h'], process.argv[2])) {
+    process.argv.splice(2, 0, 'server');
```
is it necessary to mutate process.argv ?

jlipps replied:
yes, argparse reads global process.argv. We could re-set it after running the parser to be on the safe side though.

jlipps replied:
I found a different solution that involves overriding the parseArgs method. Yay for no mutation.

jlipps commented on lib/drivers.js:
```diff
@@ -0,0 +1,157 @@
+import log from './logger';
+
+// TODO the location of the driver data file should be configurable
+const DRIVERS_JSON = path.resolve(__dirname, '..', '..', '.drivers.json');
```
I was thinking about this. If someone reinstalls appium, then all the drivers need to also be reinstalled. But if .drivers.json lives in the home folder, it will be out of sync. Thoughts on this design?

jlipps replied:
When the Appium server starts, should it verify that what's in .drivers.json matches what's installed in node_modules? We could then have two new driver commands which are mentioned in an error message:

```
appium driver verify  # prints out errors if anything in .drivers.json is not actually installed
appium driver reinstall  # reinstalls any that don't exist
```

On this kind of model it would make sense to have .drivers.json in $HOME, maybe should be called `.appium-drivers.json`, or we have a dir `.appium/drivers.json`, in case we need more stuff there in the future.

mykola-mokhnach replied:
in case with Appium Desktop we usually cannot write into its installation folder

jlipps replied:
Yeah, good point. We should have a `.appium` dir which can host `.appium/drivers.json` and `.appium/plugins.json` and maybe the driver npm modules themselves. (see forthcoming comment)

jlipps replied:
@mykola-mokhnach I had the thought that we could just keep all the installed modules in a location separate from appium, and then simply `require` that full path rather than the bare module name. This is also nice because when people reinstall Appium, their drivers don't get blown away.

jlipps replied:
This seems to work well in practice!

jlipps commented:
Another design question: currently, there are a lot of appium server flags that really only pertain to specific drivers. For example `chromedriverExecutablePath` is not relevant to the other drivers.

For some of these, it's ok to just remove them, because users can set them via desired capabilities instead.

But some of them would be a security risk as a desired cap (like being able to execute any binary on the system with the cap above), so it should really be set when the server is started by the system admin. Here are some options I've thought of for how to deal with this:

1. add a new server arg `--driver-args` that allows a JSON object or file to be defined, something like `{"uiautomator2": {"chromedriverExecutablePath": "/usr/bin/chromedriver"}}`. Then Appium will pass these args along if that driver is instantiated.
2. Each driver could export its own custom argparse args, and when Appium starts up, it first checks installed drivers, and retrieves their custom server args, and merges them in with its own args. So `--chromedriverExecutablePath` would still be available as a server arg for anyone who has installed the uiautomator2 driver.

Other ideas?

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+  }
+}
+
+function startSpinner (msg) {
```
I'd rather put these helpers into separate module

jlipps replied:
👍

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+    }
+  }
+
+  stopSpinner();
```
should this be called in finally block?

jlipps replied:
ah yes, good point.

jlipps replied:
in practice, every error thrown ends up halting the process, so i think it's fine. but finally is much safer/cleaner.

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+      drivers[name].updateVersion = null;
+      if (installType === INSTALL_TYPE_NPM) {
+        const latestVersion = await getLatestVersion(pkgName);
+        if (!version || semver.gt(latestVersion, version)) {
```
use appium-support.util.compareVersions

jlipps replied:
👍

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+
+  const drivers = {};
+
+  for (const name of Object.keys({...INSTALLED_DRIVERS})) {
```
what is the purpose of `{...INSTALLED_DRIVERS}` ?

jlipps replied:
I thought I was protecting against mutation of INSTALLED_DRIVERS. but i wasn't.

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+  return startSpinner(msg);
+}
+
+async function list ({json, showInstalled, showUpdates}) {
```
I would say we need file-based locking mechanism to prevent parallel driver installation, like npm or apt does

jlipps replied:
you mean lock the drivers.json file during operations?

mykola-mokhnach replied:
not only that. There might be a mess when two parallel appium instances are trying to preform changes to the npm registry

jlipps replied:
@mykola-mokhnach how would you go about adding the lock?

jlipps replied:
I used https://github.com/appium/appium-support/pull/179 to address this.

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+    installType = INSTALL_TYPE_LOCAL;
+    const stopSpinner = logSpin(json, 'Linking driver from local path');
+    const pkgJsonData = await linkPackage(installSpec);
+    stopSpinner();
```
finally

mykola-mokhnach replied:
same below

mykola-mokhnach commented on lib/npm.js:
```diff
@@ -0,0 +1,102 @@
+    args.push('-json');
+  }
+  execOpts.cwd = APPIUM_ROOT;
+  const {stdout, stderr, code} = await exec('npm', args, execOpts);
```
npm is a batch script on windows, so this won't work there

jlipps replied:
i'm able to run commands on my windows machine just fine using `npm`

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,198 @@
+
+async function writeInstalledDriver (name, data) {
+  INSTALLED_DRIVERS[name] = data;
+  await fs.writeFile(DRIVERS_JSON, JSON.stringify(INSTALLED_DRIVERS));
```
should we also set encoding explicitly?

jlipps replied:
👍

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,198 @@
+  }
+
+  log.info('Available drivers:');
+  for (const driverName of Object.keys(INSTALLED_DRIVERS)) {
```
why not deconstruct directly here?

jlipps replied:
`Object.keys(INSTALLED_DRIVERS)` is an array, not an object.

mykola-mokhnach replied:
I mean you can iterate `_.toPairs()` and then deconstruct it to `[driverName, {version, automationName}]`

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,198 @@
+}
+
+function getDriverBySupport (automationName, platformName) {
+  for (const driverName of Object.keys(INSTALLED_DRIVERS)) {
```
why not deconstruct here?

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,198 @@
+  for (const driverName of Object.keys(INSTALLED_DRIVERS)) {
+    const platformNames = INSTALLED_DRIVERS[driverName].platformNames;
+    if (!_.isArray(platformNames)) {
+      msg += `Entry for '${driverName}' has missing or ` +
```
I would also print the actual values in the error message

jlipps replied:
going to refactor this whole bit so it's cleaner and can allow validation to proceed when some drivers are valid and some aren't.

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,198 @@
+  try {
+    INSTALLED_DRIVERS = require(DRIVERS_JSON);
+  } catch (err) {
+    if (existsSync(DRIVERS_JSON)) {
```
what is the reason to not use the async implementation?

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+  !json && console.log(msg);
+}
+
+async function runDriverCommand (args) {
```
I would be happy if functions could have docstrings with parameters and returns types descriptions

jlipps replied:
I promise to add them once the actual methods and parameters are approved.

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+  return INSTALLED_DRIVERS;
+}
+
+async function installViaNpm (json, driver, pkgName, pkgVer) {
```
pkgVer can be undefined though

mykola-mokhnach replied:
also, this function has too many args. Consider packing them into an object

mykola-mokhnach commented on lib/npm.js:
```diff
@@ -0,0 +1,102 @@
+  return (await npmExec('view', [pkg, 'dist-tags'], {json: true})).json.latest;
+}
+
+async function installPackage (pkg, ver) {
```
ver can be null/undefined

jlipps replied:
that's ok, it's checked in callee.

mykola-mokhnach commented on lib/npm.js:
```diff
@@ -0,0 +1,102 @@
+    // if we tried to install a package via git/github, we might not have been
+    // certain of the package name, so if we can retrieve it unambiguously from
+    // the json report of the install, do so
+    const names = _.uniq(res.json.added.concat(res.json.updated).map(x => x.name));
```
use `[...arr1, ...arr2]` format for concatenation

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+}
+
+function logSpin (json, msg) {
+  if (json) {
```
ternary operator would be shorter

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+  }
+
+  stopSpinner();
+  if (!json) {
```
if json return drivers

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+
+  if (!driverData.driverName || !driverData.version || !driverData.automationName ||
+      !driverData.platformNames || !driverData.mainClass) {
+    throw new Error(`Installed driver did not expose correct fields for ` +
```
I'd rather explicitly say which field(s) are missing

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+  if (!pkgName) {
+    pkgName = driver;
+  }
+  const npmSpec = `${pkgName}${pkgVer ? '@' + pkgVer : ''}`;
```
npm also supports special identifiers, like `latest` or `beta`. Are they properly handled?

jlipps replied:
Yes, any npm identifier will work here. We just pass it straight from the command line.

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+  const stopSpinner = logSpin(json, msg);
+  try {
+    const pkgJsonData = await installPackage(pkgName, pkgVer);
+    stopSpinner();
```
should be in finally

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,265 @@
+async function runDriverCommand (args) {
+  let jsonResult = null;
+  try {
+    INSTALLED_DRIVERS = validateDriversJson();
```
we probably would need some command, like reset, which is going to recreate/resync drivers json if it was damaged/edited manually. With the current approach there is no way to launch Appium if this config is damaged

jlipps replied:
i've added a todo for this, my thought was as follows:

```
# verify driver.json and that every driver referred to there can be required OK
# just print out the status
appium driver verify

# perform verification but then reinstall anything which couldn't be required
appium driver verify --reinstall
```

mykola-mokhnach replied:
would it fix a problem with damaged config?

jlipps replied:
what kind of config problems could be fixable automatically? here are the kinds of problems i can think of:

1. driver is mentioned in config but is not npm installed. this case would be fixed with `--reinstall`
2. driver is npm installed but not mentioned in config. the only way to fix this automatically would be to require the package.json of every module in the appium home dir, and check for appium driver metadata. this seems ... slow. also, what if it was npm linked in instead of installed? or simply copied in? there's no way to guarantee that a package in node_modules can be re-installed via NPM.
3. config for driver has a missing key. how could we automatically generate the value for a missing key?
4. config for driver has an invalid value for a key. how could we decide how to fix it automatically?
5. config for driver has an additional key beyond the valid ones. this could be removed automatically, but what's the point?

Can you think of any other "damaged" cases and how we might fix them? I can't, which is why it seemed a simple reinstall of missing modules is the best we can do.

mykola-mokhnach replied:
I just wanted to make it simple at the beginning. I mean it probably would be enough to go without config for the MVP and simply use npm for packages control.

mykola-mokhnach commented on lib/main.js:
```diff
@@ -27,6 +29,7 @@ async function preflightChecks (parser, args, throwInsteadOfExit = false) {
     }
     warnNodeDeprecations();
     validateServerArgs(parser, args);
+    validateDriversJson();
```
I think this verification must be forgivable

jlipps committed:
feat: add --appium-home (alias --home, -ah) to allow setting location of appium dir
fix: install and require driver modules from appium home dir instead of appium node package

refactor: use toPairs instead of Object.keys wherever possible

refactor: move driver data validation to a better

refactor: make a server-parser file to mirror driver-parser

feat: change install source arg to --source. Also add unit tests for driver parser

feat: use yaml for driver manifest format instead of json

feat: add schema version to yaml format

mykola-mokhnach commented:
I'm not quite sure how much we need that, but usually one adds the actual schema version number into configs, so we can easily alert the schema and still handle it properly in never appium versions and/or perform config migration. This touches both drivers config and the config we add into package.json for each driver

jlipps commented:
> I'm not quite sure how much we need that, but usually one adds the actual schema version number into configs, so we can easily alter the schema and still handle it properly in newer appium versions and/or perform config migration.

A schema number is a good idea 👍

mykola-mokhnach commented:
I personally prefer YAML as main schema format, since it supports comments and is easier for manual edit

jlipps commented:
> I personally prefer YAML as main schema format, since it supports comments and is easier for manual edit

@imurchie @KazuCocoa and anyone else, what do you think? In my mind, staying with JSON was a nice and easy thing to do. I do agree YAML's a bit easier to read, though just pretty-printing the JSON would go a long way here.

KazuCocoa commented:
I'm for JSON basically.

But as Mykola commented as _comments_, we can leave comments in the YAML file.
If the file will be managed by multiple persons, such comments will be helpful.
(Nested YAML will be complicated to manage, but I think this file won't be so.)

Is this about 'driver.json', btw?
https://github.com/appium/appium/pull/13463/files#diff-726dd6ffdf869f938cc3750167ab1521R12

imurchie commented:
yaml is definitely better. Comments, partials, etc.

jlipps commented:
> Is this about 'driver.json', btw?
Yes.

jlipps commented:
> yaml is definitely better. Comments, partials, etc.

@imurchie do we use a yaml library in our codebase somewhere already?

jlipps commented:
this one looks pretty great: https://www.npmjs.com/package/yaml

jlipps commented:
OK, I switched to YAML. I like it better too.

mykola-mokhnach commented on .gitignore:
```diff
@@ -53,3 +53,4 @@ sample-code/csharp/test/obj
 sample-code/csharp/test/bin
 sample-code/csharp/packages
 appium.zip
+.drivers.json
```
is this still needed?

mykola-mokhnach replied:
isn't this in yaml format now?

mykola-mokhnach commented on lib/cli-utils.js:
```diff
@@ -0,0 +1,66 @@
+  const stopSpinner = logSpin(json, msg);
+  let res;
+  try {
+    res = await fn();
```
`return await fn()`

finally block will be properly executed after return

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,247 @@
+        break;
+      case 'verify':
+        // TODO create a 'verify' command that just checks driver.json and then
+        // that everything in driver.json can be required OK.
```
is it still .json?

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,247 @@
+
+  await spinWith(json, lsMsg, async () => {
+    if (showUpdates) {
+      for (const [, data] of _.toPairs(drivers)) {
```
you can use _.values(drivers) if keys are not needed

mykola-mokhnach replied:
`for (const {version, pkgName, installType} of _.values(drivers))`

jlipps replied:
I’m actually mutating the value of drivers though, so I need access to it.

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,247 @@
+  const specMsg = npmSpec === driver ? '' : ` using NPM install spec '${npmSpec}'`;
+  const msg = `Installing '${driver}'${specMsg}`;
+  try {
+    const pkgJsonData = await spinWith(json, msg, async () => {
```
`await spinWith(json, msg, () => installPackage(appiumHome, pkgName, pkgVer))`

jlipps replied:
Was just too long for one line.

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,247 @@
+
+  const reqFields = ['driverName', 'automationName', 'platformNames', 'mainClass'];
+  const missingFields = [];
+  for (const reqField of reqFields) {
```
I'd use reduce here

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,232 @@
+    }
+  }
+
+  throw new Error(`Could not find installed driver to support given caps`);
```
I'd rather explicitly state the caps

mykola-mokhnach commented on lib/npm.js:
```diff
@@ -0,0 +1,108 @@
+  if (!opts.appiumHome) {
+    throw new Error('appiumHome was not provided to npmExec');
+  }
+  execOpts.cwd = opts.appiumHome;
```
it is not a good practice to mutate object arguments

mykola-mokhnach commented:
I was also thinking about the following scenario:
- I install Appium uia2 driver
- Appium uia1 driver is installed as a dependency: do we show it in the installed drivers list?
- I see the list of driver and I want to uninstall/downgrade uia1 driver, because I don't use it: what do we do in such case? do we also uninstall the depending uia2 driver?
or
- I installed uia1 driver of particular version
- Now I want to install uia2 driver, but that one is not compatible with uia1 driver having this version: what would be the behaviour in such case?

jlipps committed:
fix lint warnings and use reduce

jlipps commented:
> I was also thinking about the following scenario...

Would it work to `npm install` each driver into its own subdirectory of the driver registry? That way UIA1 and UIA2 could be installed side-by-side.

jlipps committed:
fix: add file-based lock mechanism to npm installs and links

jlipps committed:
fix: install each driver into its own subdirectory
to ensure that npm install doesn't do weird things with different subdependencies of each

jlipps commented:
OK @mykola-mokhnach I think I've addressed all the concerns. Can you review the code again?

jlipps committed:
fix: set up travis for appium 2.0

mykola-mokhnach commented on lib/cli-utils.js:
```diff
@@ -0,0 +1,68 @@
+
+function errAndQuit (json, msg) {
+  if (json) {
+    console.log(JSON.stringify({error: msg.toString()}));
```
```
`${msg}` is more secure than msg.toString()
```

mykola-mokhnach commented on lib/cli-utils.js:
```diff
@@ -0,0 +1,68 @@
+  process.exit(1);
+}
+
+function log (json, msg) {
```
it would be nice to have docstrings on these methods

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,250 @@
+  }
+
+  if (json) {
+    console.log(JSON.stringify(jsonResult)); // eslint-disable-line no-console
```
should the output be prettified?

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,250 @@
+
+  const reqFields = ['driverName', 'automationName', 'platformNames', 'mainClass'];
+  const missingFields = reqFields.reduce((acc, field) => (
+    appium[field] ? acc : acc.concat(field)
```
[...acc, field]

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,250 @@
+  const {driverName, automationName, platformNames, mainClass} = appium;
+
+  const reqFields = ['driverName', 'automationName', 'platformNames', 'mainClass'];
+  const missingFields = reqFields.reduce((acc, field) => (
```
could we use lodash _.difference helper instead?

jlipps replied:
i don't think so, we're not trying to find the difference between two arrays. we're trying to find which keys from a short list don't exist on an object. there's probably a way to recast this problem in terms of _.difference but i don't think it would be clearer than reduce.

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,242 @@
+  windows: 'appium-windows-driver',
+  mac: 'appium-mac-driver',
+  espresso: 'appium-espresso-driver',
+  tizen: 'appium-tizen-driver',
```
flutter?

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,242 @@
+  tizen: 'appium-tizen-driver',
+};
+
+const INSTALL_TYPE_NPM = 'npm';
```
this could be a map

jlipps replied:
i am not sure how this would improve the situation? the reason for making it a constant is to not have to worry about the particular string value that we use in the cli to refer to 'npm' type installs.

mykola-mokhnach commented on lib/drivers.js:
```diff
@@ -0,0 +1,242 @@
+  INSTALL_TYPE_NPM
+];
+
+async function validateDriversYaml (appiumHome, logFn = log.error) {
```
I'd rather extract all the schema-related stuff to a separate class

mykola-mokhnach commented on lib/npm.js:
```diff
@@ -0,0 +1,125 @@
+const INSTALL_LOCKFILE = '.appium.install.lock';
+const LINK_LOCKFILE = '.appium.link.lock';
+
+async function npmExec (cmd, args, opts, execOpts = {}) {
```
how about putting these npm-related functions into a class?

mykola-mokhnach commented on lib/server-parser.js:
```diff
@@ -0,0 +1,377 @@
+import { DEFAULT_BASE_PATH } from 'appium-base-driver';
+
+// serverArgs will be added to the `server` (default) subcommand
+const serverArgs = [
```
I think it makes sense to put server args into a separate json file

jlipps replied:
they can't totally be in a json file since they reference js functions, but i moved them into their own file.

mykola-mokhnach commented on lib/server-parser.js:
```diff
@@ -0,0 +1,377 @@
+  }],
+];
+
+function parseSecurityFeatures (features) {
```
There is a potential issue that some third party drivers could mimic the same feature names. Perhaps, it would make sense to prefix each feature name with the corresponding automation name it affects (e.g. UiAutomator2:adb_shell)

jlipps replied:
this is a good point, however given that we can't prevent the third party drivers from just `rm -rf /` anyway, there doesn't seem to be much use in trying to create a restricted system. to even use a 3rd party driver you have to trust it totally. authors of 3rd party drivers can opt into appium's security model, just the way we do as 1st party authors. if they want to do something nefarious, they don't need to worry about breaking our security model, they can just do it.

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,250 @@
+  }
+}
+
+async function list ({appiumHome, json, showInstalled, showUpdates}) {
```
it makes sense to have the commands as members of the same class, since they share common args

mykola-mokhnach commented on lib/driver-cli.js:
```diff
@@ -0,0 +1,250 @@
+  const {json, appiumHome} = args;
+  const logFn = (msg) => log(json, msg);
+  try {
+    INSTALLED_DRIVERS = await validateDriversYaml(appiumHome, logFn);
```
I wouldn't focus on the format here, but rather on the business logic. how about changing the func name to  `validateDriversConfig` ?

jlipps commented:
this was merged on accident, but i force-pushed to the target branch, so it is now unmerged. i'm going to open a new PR for this once i've resolved the comments here.
